plugins/utils/todo-comments.nix

```
@@ -1,290 +1,452 @@
-          in
-          optional (keymap != null) {
-            mode = "n";
-            inherit (keymap) key options;
-            action = ":${funcName}${cwd}${keywords}<CR>";
-          }
-        ) commands
-      );
-
-      # Automatically enable plugins if keymaps have been set
-      plugins = mkMerge [
-        (mkIf (cfg.keymaps.todoTrouble != null) { trouble.enable = true; })
-        (mkIf (cfg.keymaps.todoTelescope != null) { telescope.enable = true; })
-      ];
+          mkCompositeOption "Keymap settings for the `:${funcName}` function." {
+            key = mkOption {
+              type = types.str;
+              default = null;
+              description = "Key for the `${funcName}` function.";
+            };
+
+            cwd = mkOption {
+              type = types.nullOr types.str;
+              description = "Specify the directory to search for comments";
+              default = null;
+              example = "~/projects/foobar";
+            };
+
+            keywords = mkOption {
+              type =
+                with types;
+                transitionType str (string: if string != null then splitString "," string else null) (
+                  nullOr (listOf str)
+                );
+              description = ''
+                Comma separated list of keywords to filter results by.
+                Keywords are case-sensitive.
+              '';
+              default = null;
+              example = "TODO,FIX";
+            };
+
+            options = keymaps.mapConfigOptions;
+          };
+      in
+      mapAttrs mkKeymapOption keymapsActions;
+
+    ripgrepPackage = mkPackageOption {
+      name = "ripgrep";
+      default = pkgs.ripgrep;
     };
+  };
+
+  extraConfig = cfg: {
+    assertions = [
+      {
+        assertion = cfg.keymaps.todoTelescope.key or null != null -> config.plugins.telescope.enable;
+        message = ''
+          Nixvim(plugins.todo-comments): You have enabled todo-comment's `telescope` integration.
+          However, you have not enabled the `telescope` plugin itself (`plugins.telescope.enable = true`).
+        '';
+      }
+      {
+        assertion = cfg.keymaps.todoTrouble or null != null -> config.plugins.trouble.enable;
+        message = ''
+          Nixvim(plugins.todo-comments): You have enabled todo-comment's `trouble` integration.
+          However, you have not enabled the `trouble` plugin itself (`plugins.trouble.enable = true`).
+        '';
+      }
+    ];
+
+    extraPackages = [ cfg.ripgrepPackage ];
+
+    keymaps = lib.pipe cfg.keymaps [
+      (filterAttrs (n: v: v != null))
+      (mapAttrsToList (
+        name: keymap: {
+          inherit (keymap) key options;
+          mode = "n";
+          action =
+            let
+              cmd = keymapsActions.${name};
+              cwd = optionalString (keymap.cwd != null) " cwd=${keymap.cwd}";
+              keywords = optionalString (
+                keymap.keywords != null
```

```suggestion
                keymap.keywords != null && keymap.keywords != [ ]
```

Or is it intentional to be able to pass an empty list?

If not, maybe we can get rid of the `nullOr` (moving it to the left of the `transitionType`)?

---

Not sure why someone would choose to pass an empty list, it should be the same effect as null to the plugin so it makes sense to add it to the condition to avoid trying to concat an empty list.